src/files/XDRFile.cpp

```diff
@@ -222,66 +249,71 @@ static void encodeints(std::vector<char>& buf, DecodeState& state, uint32_t num_
-            tmp >>= 8;
+            tmp >>= CHAR_BIT;
         }
         while (tmp != 0) {
             bytes[bytecnt++] = tmp & 0xff;
-            tmp >>= 8;
+            tmp >>= CHAR_BIT;
         }
         num_of_bytes = bytecnt;
     }
-    if (num_of_bits >= num_of_bytes * 8) {
+    // If the caller specified a sufficiently large bit count,
+    // do what they say.
+    if (num_of_bits >= num_of_bytes * CHAR_BIT) {
         for (size_t i = 0; i < num_of_bytes; i++) {
-            encodebits(buf, state, 8, bytes[i]);
+            encodebits(buf, state, CHAR_BIT, bytes[i]);
         }
-        encodebits(buf, state, num_of_bits - num_of_bytes * 8, 0);
+        encodebits(buf, state, num_of_bits - num_of_bytes * CHAR_BIT, 0);
     } else {
+        // Otherwise send each byte we found
         size_t i;
         for (i = 0; i < num_of_bytes - 1; i++) {
-            encodebits(buf, state, 8, bytes[i]);
+            encodebits(buf, state, CHAR_BIT, bytes[i]);
         }
-        encodebits(buf, state, num_of_bits - (num_of_bytes - 1) * 8, bytes[i]);
+        const uint32_t num_bits_remaining = num_of_bits - (num_of_bytes - 1) * CHAR_BIT;
+        assert(num_bits_remaining < CHAR_BIT); // Help clang analyzer understand
+        encodebits(buf, state, num_bits_remaining, bytes[i]);
     }
 }
 
 /*
  * decodebits - decode number from buf using specified number of bits
  *
- * Extract the number of bits from the array buf and construct an integer
+ * Equivalent to `receivebits` in Gromacs.
+ *
+ * Extract the number of bits from the databuffer and construct an integer
  * from it. Return that value.
  * The number of bits must be <32.
  *
  */
 
 template <typename T>
 static T decodebits(const std::vector<char>& buf, DecodeState& state, uint32_t num_of_bits) {
-    uint32_t mask = static_cast<uint32_t>(1 << num_of_bits) - 1;
+    const uint32_t mask = static_cast<uint32_t>(1 << num_of_bits) - 1;
 
-    size_t cnt = state.count;
     size_t lastbits = state.lastbits;
     uint32_t lastbyte = state.lastbyte;
 
     uint32_t num = 0;
-    while (num_of_bits >= 8) {
-        lastbyte = (lastbyte << 8) | static_cast<uint8_t>(buf[cnt++]);
-        num |= (lastbyte >> lastbits) << (num_of_bits - 8);
-        num_of_bits -= 8;
+    while (num_of_bits >= CHAR_BIT) {
+        lastbyte = (lastbyte << CHAR_BIT) | static_cast<uint8_t>(buf[state.count++]);
+        num |= (lastbyte >> lastbits) << (num_of_bits - CHAR_BIT);
+        num_of_bits -= CHAR_BIT;
     }
     if (num_of_bits > 0) {
         if (lastbits < num_of_bits) {
-            lastbits += 8;
-            lastbyte = (lastbyte << 8) | static_cast<uint8_t>(buf[cnt++]);
+            lastbits += CHAR_BIT;
+            lastbyte = (lastbyte << CHAR_BIT) | static_cast<uint8_t>(buf[state.count++]);
         }
         lastbits -= num_of_bits;
-        num |= (lastbyte >> lastbits) & mask;
+        num |= (lastbyte >> lastbits) & (static_cast<uint32_t>(1 << num_of_bits) - 1);
```

Comment:
I think this was an error which I introduced when writing the code. Sorry!
However, this seemingly does not effect any real world data because I've tested with some longer trajectories and this gives same output when reading.
And, tests pass anyways.